### PR TITLE
Fix animated image playback bugs

### PR DESCRIPTION
**Pull request by ZXIOU**
Hello, I'm an engineer from Xiaomi. In the actual development, I found that some animated image are not displayed correctly, such as the following GIF picture：[https://s2.ax1x.com/2019/11/01/KHYIgJ.gif](https://s2.ax1x.com/2019/11/01/KHYIgJ.gif) 。

I read about the implementation of **sdaimatedimageview** and **sdaimatedimageplayer** in detail, and found that the **currentFrame** displayed and **currentFrameIndex** do not match, which will lead to wrong frame playback. Although the playback time of each frame is correct, but the rendered picture is the picture of the previous frame. The reason for the good performance of the examples in demo is that the time of each frame is the same or the difference is very small, so there is no difference in vision. But if the time gap between frames becomes larger, it will cause problems.

WX20191121-200851@2x

WX20191121-200933@2x

From the above figure, it can be seen that the **currentframe** and **currentframeindex** are not consistent, but in the next **displaydidrefresh** call, we have no chance to display the pictures corresponding to the latest index, resulting in that the pictures displayed in frame N + 1 are always the pictures of frame N. So at the beginning, you have to check if you want to show new pictures.

And I also found that we can't set **animationrepeatcount** after calling **setImage**.

I added the display of the problem pictures in the project demo. Because this picture setting is only displayed once, so now I have manually set unlimited playback for this picture. You can test and verify it.

### CODE REVIEW

**Comment by dreampiggy** [comment]
Hi. Thanks for your feedback and PR.

Does this bug behavior exist only from the SDWebImage [v5.3.0](https://github.com/SDWebImage/SDWebImage/releases/tag/5.3.0) ? Or it exists even from 5.0.0-5.2.0 ?

Because only after 5.3.0, we do refactory to seperate the Animated Image View and Player, to make it suitable for more concepts (like the case you don't have any CALayer).

**Comment by ZXIOU** [comment]
> Hi. Thanks for your feedback and PR.
> 
> Does this bug behavior exist only from the SDWebImage [v5.3.0](https://github.com/SDWebImage/SDWebImage/releases/tag/5.3.0) ? Or it exists even from 5.0.0-5.2.0 ?
> 
> Because only after 5.3.0, we do refactory to seperate the Animated Image View and Player, to make it suitable for more concepts (like the case you don't have any CALayer).

This problem has existed since we officially abandoned **FLAnimatedfImage**. I think it should be 5.0.0. I found that after recently refactoring with SDAnimatedImagePlayer, the core render code has not changed, so this problem has always existed.

File: Examples/SDWebImage Demo/DetailViewController.m [hunk 1]
```diff
@@ -24,6 +24,8 @@ - (void)configureView {
     [self.imageView sd_setImageWithURL:self.imageURL
                       placeholderImage:nil
                                options:SDWebImageProgressiveLoad];
+    self.imageView.shouldCustomLoopCount = YES;
+    self.imageView.animationRepeatCount = NSIntegerMax;
```
Comment (dreampiggy) [inline Examples/SDWebImage Demo/DetailViewController.m]: If you want infinity loop, actually, use 0 is OK :)

File: SDWebImage/Core/SDAnimatedImageView.m [hunk 2]
```diff
@@ -306,6 +306,19 @@ - (void)setHidden:(BOOL)hidden
...
 
+- (void)setAnimationRepeatCount:(NSInteger)animationRepeatCount
+{
+#if SD_UIKIT
+    [super setAnimationRepeatCount:animationRepeatCount];
+#else
+    _animationRepeatCount = animationRepeatCount;
+#endif
+    
+    if (self.shouldCustomLoopCount) {
```
Comment (dreampiggy) [inline SDWebImage/Core/SDAnimatedImageView.m]: Good catch. Another issue during the refactory. Maybe v5.3.2 is soon.

Comment (dreampiggy) [reply SDWebImage/Core/SDAnimatedImageView.m]: Do I need to add one more test case about this ? Avoid future regression issue.

```objective-c
SDAnimatedImageView *imageView;
imageView.image = animatedImage;
[imageView startAnimating];
imageView.shouldCustomLoopCount = YES;
imageView.animationRepeatCount = 100;
expect(imageView.player.totalLoopCount).equal(100);
```

File: Examples/SDWebImage Demo/DetailViewController.m (hunk 1 above)
Comment (ZXIOU) [reply Examples/SDWebImage Demo/DetailViewController.m]: OOKK, I just missed it xdddddd

File: SDWebImage/Core/SDAnimatedImageView.m (hunk 2 above)
Comment (ZXIOU) [reply SDWebImage/Core/SDAnimatedImageView.m]: I think it's enough to set the animationrepeatcount property. But if we want to do better, when setting the new animationrepeatcount property, we should compare the size of the old and new values, and judge the current playing state. Theoretically, there should be six situations, but there may be only two that need to be dealt with, that is, the current playback is not over, when the new value is small (including unlimited playback), and when the current playback times are greater than the new value, the playback will be stopped immediately; the other is when the current playback is completed and the playback times are less than the new value, the playback will continue. This code is not RP, but I want to discuss whether to add this logic.

Comment (ZXIOU) [reply SDWebImage/Core/SDAnimatedImageView.m]: > Do I need to add one more test case about this ? Avoid future regression issue.
> 
> ```objc
> SDAnimatedImageView *imageView;
> imageView.image = animatedImage;
> [imageView startAnimating];
> expect(imageView.player.totalLoopCount).equal(0);
> imageView.shouldCustomLoopCount = YES;
> imageView.animationRepeatCount = 100;
> expect(imageView.player.totalLoopCount).equal(100);
> ```

I think it's enough to set the animationrepeatcount property. But if we want to do better, when setting the new animationrepeatcount property, we should compare the size of the old and new values, and judge the current playing state. Theoretically, there should be six situations, but there may be only two that need to be dealt with, that is, the current playback is not over, when the new value is small (including unlimited playback), and when the current playback times are greater than the new value, the playback will be stopped immediately; the other is when the current playback is completed and the playback times are less than the new value, the playback will continue. This code is not RP, but I want to discuss whether to add this logic.

File: Examples/SDWebImage Demo/DetailViewController.m (hunk 1 above)
Comment (ZXIOU) [reply Examples/SDWebImage Demo/DetailViewController.m]: > If you want infinity loop, actually, use 0 is OK :)

OOKK, I just missed it xdddddd

File: SDWebImage/Core/SDAnimatedImageView.m (hunk 2 above)
Comment (dreampiggy) [reply SDWebImage/Core/SDAnimatedImageView.m]: Maybe not. This is a fix version, all previous behavior, just change the loop count, does not do anything related to the stop/start logic. Let's just keep the logic the same as before.

This test case can ensure `when setup the Animated ImageView.loopCount, we indeed update the player loop count`

**Comment by dreampiggy** [comment]
Sometimes, the CI test contains issue. Restart does not solve the issue.
You need to a force push (ammend a empty commit) to make it works. I don't know the reason now.

**Comment by dreampiggy** [comment]
LGTM. I can merge and add the test case for this. Any idea ?

**Comment by ZXIOU** [comment]
OK, I see. Thank you.

**Comment by ZXIOU** [comment]
If you think it's OK, I think you can do that right away. I totally agree.

**Comment by dreampiggy** [comment]
@ZXIOU Released [v5.3.2](https://github.com/SDWebImage/SDWebImage/releases/tag/5.3.2). You can use that version if needed.

**Comment by ZXIOU** [comment]
OK, thank you :)

**Comment by dreampiggy** [comment]
@ZXIOU  This PR break the progressive animation decoding, including GIF/APNG/FLIF, etc.

The lastFrameIndex was set to the wrong number, which cause the proressive index is wrong.

Need a patch for both 5.3.x, 5.4.x, 5.5.x..

**Comment by dreampiggy** [comment]
See video.

[Progressive Animation Bug.zip](https://github.com/SDWebImage/SDWebImage/files/4080088/Progressive.Animation.Bug.zip)

**Pull request merged**